### PR TITLE
MOD-6540: Support EMPTY indexing for TEXT fields (#4622)

### PR DESCRIPTION
**Pull request by raz-mon**
**Describe the changes in the pull request**

Adds support for `EMPTY TEXT` indexing - please refer to #4622's comment for more information.

Also:
* Removes support of `EMPTY TAG` indexing for empty JSON arrays and objects.
* Cleans up a bunch of code (dead mainly).

**Which issues this PR fixes**
1. MOD-6580.
2. MOD-6540

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

**Review by oshadmi — COMMENTED** [review]
Need to update `FT.CREATE` in commands.json with `ISEMPTY` argument?
(should also be aligned with documentation PR https://github.com/redis/docs/pull/199)

**Comment by raz-mon** [comment]
@oshadmi This is already done, in the very PR you attached :)

**Comment by fcostaoliveira** [comment]
### Automated performance analysis summary

This comment was automatically generated given there is performance data available.

In summary:
- Detected a total of 12 stable tests between versions.
- Detected a total of 2 highly unstable benchmarks.
- Detected a total of 4 improvements above the improvement water line.
- Detected a total of 2 regressions bellow the regression water line 5.0.

You can check a comparison in detail via the [grafana link](https://benchmarksrediscom.grafana.net/d//3Ejv2wZnk/?var-branch=master&var-branch=razmon-feature_empty_text_indexing)

### Comparison between master and razmon-feature_empty_text_indexing.

Time Period from 30 days ago. (environment used: oss-standalone)

|                                        Test Case                                        |Baseline master (median obs. +- std.dev)|Comparison razmon-feature_empty_text_indexing (median obs. +- std.dev)|% change (higher-better)|                Note                |
|-----------------------------------------------------------------------------------------|----------------------------------------|---------------------------------------------------------------------:|------------------------|------------------------------------|
|ftsb-10K-enwiki_abstract-hashes-fulltext-sortby                                          | 68 +- 0.8% (7 datapoints)              |                                                                 68.00|0.4%                    |No Change                           |
|ftsb-10K-enwiki_abstract-hashes-term-suffix                                              | 2028 +- 3.9% (7 datapoints)            |                                                               2157.00|6.4%                    |IMPROVEMENT                         |
|ftsb-10K-enwiki_abstract-hashes-term-wildcard                                            | 11959 +- 2.6% (7 datapoints)           |                                                              12295.00|2.8%                    |No Change                           |
|ftsb-1M-enwiki_abstract-hashes-fulltext-2word-intersection-query-non-sortable            | 41 +- 5.4% (7 datapoints)              |                                                                 44.00|8.6%                    |waterline=5.4%. IMPROVEMENT         |
|ftsb-1M-enwiki_abstract-hashes-fulltext-2word-intersection-query@100_ops_sec             | 100 +- 0.0% (7 datapoints)             |                                                                100.00|0.0%                    |No Change                           |
|ftsb-1M-enwiki_abstract-hashes-fulltext-2word-union-query                                | 2666 +- 2.7% (7 datapoints)            |                                                               2304.00|-13.6%                  |REGRESSION                          |
|ftsb-1M-enwiki_abstract-hashes-fulltext-2word-union-query-non-sortable                   | 1015 +- 11.5% UNSTABLE (7 datapoints)  |                                                               1115.00|9.8%                    |UNSTABLE (very high variance)       |
|ftsb-1M-enwiki_abstract-hashes-fulltext-simple-1word-query@100_ops_sec                   | 100 +- 0.0% (7 datapoints)             |                                                                100.00|0.0%                    |No Change                           |
|search-filtering-tag-numeric                                                             | 184 +- 6.4% (7 datapoints)             |                                                                211.00|14.7%                   |waterline=6.4%. IMPROVEMENT         |
|search-ftsb-10K-enwiki_abstract-hashes-term-withoutsuffix-trie                           | 40375 +- 3.0% (7 datapoints)           |                                                              40024.00|-0.9%                   |No Change                           |
|search-ftsb-1M-enwiki_abstract-hashes-fulltext-2word-union-query-non-sortable@100_ops_sec| 100 +- 0.0% (7 datapoints)             |                                                                100.00|0.0%                    |No Change                           |
|search-ftsb-1M-enwiki_abstract-hashes-fulltext-simple-1word-query-non-sortable           | 151 +- 4.2% (7 datapoints)             |                                                                147.00|-2.3%                   |No Change                           |
|search-ftsb-370K-docs-union-iterators-q4                                                 | 6.7 +- 0.8% (7 datapoints)             |                                                                  6.70|-0.6%                   |No Change                           |
|search-ftsb-5500K-docs-union-iterators-q2                                                | 0.93                                   |                                                                  0.92|-1.1%                   |No Change                           |
|search-numeric                                                                           | 3563 +- 28.8% UNSTABLE (7 datapoints)  |                                                               2791.00|-21.6%                  |UNSTABLE (very high variance)       |
|search-numeric-optimize                                                                  | 9058 +- 1.7% (7 datapoints)            |                                                               8790.00|-3.0%                   |No Change                           |
|search-numeric-sortby-optimize                                                           | 19 +- 1.9% (4 datapoints)              |                                                                 20.00|4.6%                    |potential IMPROVEMENT               |
|vecsim-arxiv-titles-384-angular-filters-m16-ef-128-fulltext-filter                       | 474 +- 3.7% (7 datapoints)             |                                                                433.00|-8.6%                   |REGRESSION                          |
|vecsim-arxiv-titles-384-angular-filters-m16-ef-128-numeric-filter                        | 105 +- 8.1% (7 datapoints)             |                                                                100.00|-5.2%                   |waterline=8.1%. potential REGRESSION|
|vecsim-arxiv-titles-384-angular-filters-m16-ef-128-tag-filter                            | 54957 +- 4.5% (7 datapoints)           |                                                              59854.00|8.9%                    |IMPROVEMENT                         |

**Comment by raz-mon** [comment]
@oshadmi Not sure, who uses the `commands.json` file?

File: tests/pytests/test_empty.py [hunk 1]
```diff
@@ -0,0 +1,641 @@
...
+
+        # Non-empty union with another field
+        conn.execute_command('HSET', 'h2', 'text', 'love you', 't', 'movie')
+        cmd = f'FT.SEARCH {idx}'.split(' ') + ['love | isempty(@t)', 'SORTBY', 'text', 'ASC']
+        expected = [
+            2, 'h1', ['text', 'hello', 't', ''], 'h2', ['text', 'love you', 't', 'movie']
+        ]
+        cmd_assert(env, cmd, expected)
+
+        if dialect == 2:
```
Comment (oshadmi) [inline tests/pytests/test_empty.py:150]: Need to be removed after fixing [MOD-6967](https://redislabs.atlassian.net/browse/MOD-6967?atlOrigin=eyJpIjoiOTQ2NjdkZjRlYThhNGUzZTk5YTRiNTFlNGM0Y2FiNzIiLCJwIjoiamlyYS1zbGFjay1pbnQifQ) (`FT.AGGREGATE` fails if apply and sortby with dialect > 2)

**Comment by oshadmi** [comment]
Reviewing benchmark: vecsim-arxiv-titles-384-angular-filters-m16-ef-128-fulltext-filter shows -8.6% but since it is unstable after consulting with @fcostaoliveira, we can proceed.

**Review by oshadmi — APPROVED** [review]

**Pull request merged**